x-pack/plugins/enterprise_search/public/applications/app_search/components/relevance_tuning/utils.ts

```diff
@@ -61,3 +61,50 @@ export const normalizeBoostValues = (boosts: Record<string, RawBoost[]>): Record
       [fieldName]: boostList.map(normalizeBoostValue),
     };
   }, {});
+
+// Our model allows for empty values to be added to boosts. However, the server will not accept
+// empty strings in values. To avoid that, we filter out empty values before sending them to the server.
+
+// I.e., the server will not accept any of the following, so we need to filter them out
+// value: undefined
+// value: []
+// value: ['']
+// value: ['foo', '']
+export const removeEmptyValueBoosts = (
+  boosts: Record<string, Boost[]>
+): Record<string, Boost[]> => {
+  // before:
+  //   { foo: { values: ['a', '', '   '] } }
+  // after:
+  //   { foo: { values: ['a'] } }
+  const trimBoostValues = (fieldBoosts: Boost[]) => {
+    return fieldBoosts.map((boost: Boost) => {
+      if (boost.type !== BoostType.Value) return boost;
+      const valueBoost = boost as ValueBoost;
+      return {
+        ...boost,
+        value: valueBoost.value.filter((v) => v.trim() !== ''),
+      };
+    });
+  };
+
+  // before:
+  //   { foo: { values: ['a'] } }
+  //   { foo: { values: [] } }
+  // after:
+  //   { foo: { values: ['a'] } }
+  const filterEmptyValueBoosts = (fieldBoosts: Boost[]) => {
+    return fieldBoosts.filter(
+      (boost) => !(boost.type === BoostType.Value && boost.value && boost.value.length === 0)
+    );
+  };
+
+  return Object.entries(boosts).reduce((acc, [fieldName, fieldBoosts]) => {
+    const updatedBoosts = filterEmptyValueBoosts(trimBoostValues(fieldBoosts));
+    if (isEmpty(updatedBoosts)) return acc;
```

Comment:
[microperf] We're essentially iterating over the same array 3 times here (`filterEmptyValueBoosts`, `trimBoostValues`, and then `isEmpty`) - this feels like it might not scale well and potentially be bad performance-wise depending on the number of value boosts on the page. Is there any way for us to combine this into a single helper that iterates only once, and returns false/null which is a much quicker check than `isEmpty`?

I'm also not a super huge fan of the use of a reducer here, it feels a little difficult to read/short-circuit (which I think we should short-circuit if it's not a value boost). I might poke at this a bit to see if I can suggest an alternative way of writing this and get your thoughts.

Reply:
Also after having typed this all out, I realize worrying about O(3n) perf is a little bit shutting the barn door after the horse has been let out considering the # of cloneDeeps on the page already 😅 

I think my concern here is more one of principle and wanting us to generally consider performance vs readability. If my forthcoming suggestion on this feels worse to you than your current code, I'm fine with sticking with this current iteration

Reply:
Ok, here's my perf refactor suggestion! It gets O(3n) down to just O(n): https://github.com/jasonstoltz/kibana/compare/rel-tuning-fixes...constancecchen:jason/rel-tuning-fixes

It may be easier to read standalone, if so: https://github.com/constancecchen/kibana/blob/b85472ab7fa8a9e805780d0f47aac60b603e34f7/x-pack/plugins/enterprise_search/public/applications/app_search/components/relevance_tuning/utils.ts#L76-L100

I tried to keep it pretty easy to read still but if you think it's still a downgrade in terms of readability, def let me know - I'd be open to keeping your 2 separate trim/filter actions and maybe just getting rid of the 3rd `isEmpty` 🤷

Reply:
>  wanting us to generally consider performance vs readability.

What is the principle you are arguing here. Performance over readability???

Reply:
I'd also be floored if this actually caused a performance issue.

Reply:
My 2 cents is this:

- Sweating over micro perf like this, unless there is an issue, is premature optimization and likely not worth the effort.
- For a method like this with solid unit tests, I don't think the readability is all that important. I'm happy to merge your changes.

Reply:
Thanks for the update Constance.

Generally speaking, I don't think it's worth the effort to sweat over micro perf like this. This feels a bit like premature optimization.

Your refactor looks fairly elegant with less code so I'm happy to merge it. TY.

Reply:
Cherry-picked your suggestion, thanks again: 0baff9a4889208886fe3bd0b89083282030eda67

Reply:
OK - definitely heard re: it being a premature optimization & not sweating microperf. It's a hard habit for me to shake as someone who used to work on a component library that had to worry about any unknown number of items on the page, as well as impacting the performance of other teams. To me it's a fun thought exercise and a way of knocking the rust off a perf mindset, but I'll try to tone it down / stress the optionality of my comments.

Reply:
I agree, thank you for pointing it out and taking the time to explain.

Reply:
> It's a hard habit for me to shake

FWIW, it's not a bad habit! Keep pointing stuff like this out, I think it's great.